### DISCUSSION TITLE
Winlibs raylib.dll.rc.data linker error

### DISCUSSION THREAD

**Discussion by Axle-Ozz-i-sofT** [raysan5/raylib#2560] [General]
Hi

I am attempting to compile raylib with FB current version "winlibs-mingw-w64-x86_64-9.3.0-7.0.0-r3-sjlj" and getting a linker error for the rc.data file. Wondering if anyone has encountered this or know how to correct it.
I have tried adding the CFLAG and LDFLAG -m64 but I still get the error. Any thoughts?

`C:\Dev-FreeBASIC\Winlibs\winlibs-mingw-w64-x86_64-9.3.0-7.0.0-r3-sjlj\raylib\src>mingw32-make V=1 -f Makefile PLATFORM=P
LATFORM_DESKTOP RAYLIB_MODULE_RAYGUI=TRUE RAYLIB_LIBTYPE=SHARED GRAPHICS=GRAPHICS_API_OPENGL_33
gcc -shared -o ../src/raylib.dll rcore.o rshapes.o rtextures.o rtext.o utils.o rglfw.o rmodels.o raudio.o raygui.o ./ray
lib.dll.rc.data  -static-libgcc -lopengl32 -lgdi32 -lwinmm -Wl,--out-implib,../src/libraylibdll.a
c:/dev-freebasic/winlibs/winlibs-mingw-w64-x86_64-9.3.0-7.0.0-r3-sjlj/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/9.3.0/..
/../../../x86_64-w64-mingw32/bin/ld.exe: i386 architecture of input file `./raylib.dll.rc.data' is incompatible with i38
6:x86-64 output
collect2.exe: error: ld returned 1 exit status
mingw32-make: *** [Makefile:489: raylib] Error 1`

Axle

**Comment by raysan5** [comment]
`raylib.dll.rc.data` is the resource file containing the icon and metadata for the `raylib.dll`, default provided version is compiled for 32bit.

You are recompiling raylib using a 64bit compiler, so, you should recompile `raylib.dll.rc` for 64bit (to generate `raylib.dll.rc.data`). [Here there are some instructions](https://github.com/raysan5/raylib/wiki/Frequently-Asked-Questions#how-do-i-setup-a-custom-icon-for-my-executable).

**Reply by Axle-Ozz-i-sofT** [reply]
Thank you for that.

Curious, I am compiling it with w64devkit you supply as a binary installer for raylib and it compiles without any issues. I thought GCC MinGW-w64 defaults to x86-64 on a 64bit machine. Seams odd as everything really defaults to 64bit in recent times?

Axle

**Reply by Axle-Ozz-i-sofT** [reply]
P.S. I just checked "https://github.com/raysan5/raylib/releases/download/4.0.0/raylib_installer_v4.0.mingw.exe" defaults to 32 bit. I didn't expect that :(

So, I guess I will need to add the 64bit flags for the resource compiler etc as you refereed to above as well as the -m64 GCC flags for the Makefile

**Reply by Axle-Ozz-i-sofT** [reply]
> C:\raylib\raylib\src>windres raylib.dll.rc -o raylib.dll.rc.data --target=pe-x86-64
windres: Can't detect target endianness and architecture.

"raylib_installer_v4.0.mingw.exe"
Windows10 21H2 64 bit

**Reply by Axle-Ozz-i-sofT** [reply]
Hi Raysan

OK, this is all a bit weird :/
under "winlibs-mingw-w64-x86_64-9.3.0-7.0.0-r3-sjlj"

> windres raylib.dll.rc -o raylib.dll.rc.data --target=pe-x86-64
**Success**

> mingw32-make V=1 -f Makefile PLATFORM=PLATFORM_DESKTOP RAYLIB_MODULE_RAYGUI=TRUE RAYLIB_LIBTYPE=SHARED GRAPHICS=GRAPHICS_API_OPENGL_33
**Success**

raylib.dll
PE  d† (aka 64 bit)

Why is "raylib_installer_v4.0.mingw.exe" [w64devkit] defaulting/forcing to 32-bit [Mingw-w64 GCC][w64] : compilers, linker, assembler
Any thoughts? (I am thinking if the default "raylib.dll.rc.data" has been compiled as PE x86 then there is a switch that I am missing?)

Alternatively I will ditch the w64devkit and go back to TDM-Dragon builds or Winlibs builds...

Axle

**Reply by raysan5** [reply]
@Axle-Ozz-i-sofT MinGW-w64 is a compiler, it can be 32bit or 64bit, there is nothing weird. raylib includes w64devkit 32bit that includes MinGW-w64 32bit. You can also download w64devkit 64bit that includes MinGW-w64 64bit. No magic in there.

It's unfortunate the name of MinGW-w64 because it can make you think it refers to 64bit but the compiler can be 32bit or 64bit, depending on the version you downloaded.

**Comment by Axle-Ozz-i-sofT** [comment]
Hi raysan5

I gave up on [raylib_installer_v4.0.mingw.exe](https://github.com/raysan5/raylib/releases/download/4.0.0/raylib_installer_v4.0.mingw.exe).
Now using Winlibs. I am using "winlibs-mingw-w64-x86_64-9.3.0-7.0.0-r3-sjlj.7z" for FB V1.09.0 compatibility.

Folder structure:

- C:\Winlibs\
    - \raylib-4.0.0\
    - \winlibs-mingw-w64-x86_64-9.3.0-7.0.0-r3-sjlj\mingw64\bin
Copy .\raygui-3.1\raygui.h to .\raylib-4.0.0\src\extras\raygui.h (Overwrite)

Batch launch the Makefile:
_REM Makefile laucher for 64-bit Compile.
REM Set the compiler and source path as required.
SET PATH=%PATH%;SET PATH=%PATH%;C:\Dev-FreeBASIC\Winlibs\winlibs-mingw-w64-x86_64-9.3.0-7.0.0-r3-sjlj\mingw64\bin
SET PATH=%PATH%;C:\Dev-FreeBASIC\Winlibs\raylib-4.0.0\src
REM Recompile the Resource file for Windows x86-64
windres raylib.dll.rc -o raylib.dll.rc.data --target=pe-x86-64
REM Using OpenGL V2.1 for VirtualBox compatability.
REM mingw32-make V=1 -f Makefile PLATFORM=PLATFORM_DESKTOP RAYLIB_MODULE_RAYGUI=TRUE RAYLIB_LIBTYPE=SHARED GRAPHICS=GRAPHICS_API_OPENGL_33
mingw32-make V=1 -f Makefile PLATFORM=PLATFORM_DESKTOP RAYLIB_MODULE_RAYGUI=TRUE RAYLIB_LIBTYPE=SHARED GRAPHICS=GRAPHICS_API_OPENGL_21
pause_

Everything compiles as expected. .\src\extras\raylib.h is found and converted to raygui.c and copied to .\src\raylib.c
So no need to manually copy/create raylib.c, and no need to copy .\src\extras\*.h to .\src

So all I can think of at the moment is some issue with raylib_installer_v4.0.mingw.exe, w64devkit and windows10 not playing nice with paths
¯\\_(ツ)_/¯ ... Win10 can be a bit weird like that.
(P.S. I did check directory and user permissions for R/W)

I haven't yet tested this under the later Winlibs versions with DW2 or SEH yet.

I just posted this up in case anyone else reads the thread and can see what I did to solve the issue, with your guidance.
Much thanks
Axle
